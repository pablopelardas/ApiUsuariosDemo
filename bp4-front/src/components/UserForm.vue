<script setup lang="ts">
import InputField from '@/components/form-controls/InputField.vue'
import { reactive, ref, watch } from 'vue'

const props = defineProps({
  user: {
    type: Object,
    required: true
  }
})

const localUser = reactive({ ...props.user })
const triedToSubmit = ref(false)

const errors = reactive<Record<string, string>>({
  names: '',
  lastnames: '',
  email: '',
  cuit: '',
  birthdate: '',
  residence: '',
  cellphone: ''
})

const validate = () => {
  Object.keys(errors).forEach((key) => {
    errors[key] = ''
  })
  triedToSubmit.value = true
  const requiredFields = [
    'names',
    'lastnames',
    'email',
    'cuit',
    'birthdate',
    'residence',
    'cellphone'
  ]
  requiredFields.forEach((field) => {
    if (!localUser[field]) {
      errors[field] = `El campo es requerido`
    }
  })
  // validate email
  if (localUser.email) {
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/
    if (!emailRegex.test(localUser.email)) {
      errors.email = 'El email no es válido'
    }
  }
  // validate cuit
  if (localUser.cuit) {
    const cuitRegex = /^\d{2}-\d{8}-\d$/
    if (!cuitRegex.test(localUser.cuit)) {
      errors.cuit = 'El CUIT no es válido'
    }
  }

  return !requiredFields.some((field) => errors[field])
}

watch(
  localUser,
  () => {
    if (triedToSubmit.value) validate()
  },
  { deep: true }
)

defineExpose({
  validate,
  localUser
})
</script>

<template>
  <div class="user-form">
    <InputField v-model="localUser.names" label="Nombres" id="names" :error="errors.names" />
    <InputField
      v-model="localUser.lastnames"
      label="Apellidos"
      id="lastnames"
      :error="errors.lastnames"
    />
    <InputField v-model="localUser.email" label="Email" id="email" :error="errors.email" />
    <InputField v-model="localUser.cuit" label="CUIT" id="cuit" :error="errors.cuit" />
    <InputField
      v-model="localUser.birthdate"
      label="Fecha de nacimiento"
      id="birthdate"
      :error="errors.birthdate"
    />
    <InputField
      v-model="localUser.residence"
      label="Residencia"
      id="residence"
      :error="errors.residence"
    />
    <InputField
      v-model="localUser.cellphone"
      label="Celular"
      id="cellphone"
      :error="errors.cellphone"
    />
  </div>
</template>
