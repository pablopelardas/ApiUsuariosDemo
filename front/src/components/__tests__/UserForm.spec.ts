import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import UserForm from '@/components/UserForm.vue'

const user = {
  names: 'John',
  lastnames: 'Doe',
  email: 'asd@asd.asd',
  birthdate: '1990-01-01',
  cellphone: '123456789',
  residence: 'Argentina',
  cuit: '20-40846184-8'
}

describe('UserForm', () => {
  it('should render correctly', () => {
    const wrapper = mount(UserForm)
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('should contain each user input field', () => {
    const wrapper = mount(UserForm)
    const fields = ['names', 'lastnames', 'email', 'birthdate', 'cellphone', 'residence', 'cuit']
    const wrapperFields = wrapper.findAllComponents({ name: 'InputField' })
    expect(wrapperFields.length).toBe(fields.length)
    console.log(wrapperFields)
    fields.forEach((field) => {
      const wrapperField = wrapperFields.find((wrapperField) => wrapperField.props('id') === field)
      expect(wrapperField?.exists()).toBe(true)
    })
  })
  it('should expose method to validate user and userData', () => {
    const wrapper = mount(UserForm)
    expect(wrapper.vm.validate).toBeDefined()
    expect(wrapper.vm.localUser).toBeDefined()
  })

  it('validation should return false if any field is empty ', () => {
    let wrapper = mount(UserForm, {
      props: {
        user: {
          ...user,
          names: ''
        }
      }
    })
    expect(wrapper.vm.validate()).toBe(false)

    wrapper = mount(UserForm, {
      props: {
        user
      }
    })
    expect(wrapper.vm.validate()).toBe(true)
  })
  it('validation should return false if cuit is invalid', () => {
    const wrapper = mount(UserForm, {
      props: {
        user: {
          ...user,
          cuit: 'asd'
        }
      }
    })
    expect(wrapper.vm.validate()).toBe(false)
  })

  it('should show error message if field is invalid', async () => {
    const wrapper = mount(UserForm, {
      props: {
        user: {
          ...user,
          names: ''
        }
      }
    })
    const wrapperField = wrapper.findComponent({ name: 'InputField', props: { id: 'names' } })
    expect(wrapperField.exists()).toBe(true)
    await wrapper.vm.validate()
    const error = wrapperField.getComponent({ name: 'InlineMessage' })
    expect(error.text()).toContain('El campo es requerido')
  })

  it('should remove error message if field is valid', async () => {
    const wrapper = mount(UserForm, {
      props: {
        user: {
          ...user,
          names: ''
        }
      }
    })
    const wrapperField = wrapper.findComponent({ name: 'InputField', props: { id: 'names' } })
    expect(wrapperField.exists()).toBe(true)
    await wrapper.vm.validate()
    const error = wrapperField.getComponent({ name: 'InlineMessage' })
    expect(error.text()).toContain('El campo es requerido')
    wrapperField.vm.$emit('update:modelValue', 'John')
    await wrapper.vm.$nextTick()
    expect(wrapper.vm.localUser.names).toBe('John')
    expect(wrapperField.vm.error).toBe('')
    expect(wrapperField.findComponent({ name: 'InlineMessage' }).exists()).toBe(false)
  })
})
