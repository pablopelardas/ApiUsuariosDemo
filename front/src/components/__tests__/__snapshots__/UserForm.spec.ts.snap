// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserForm > should render correctly 1`] = `
"<div class=\\"user-form\\">
  <div class=\\"field\\">
    <div class=\\"field-header\\"><label for=\\"names\\">Nombres</label>
      <!--v-if-->
    </div><input class=\\"p-inputtext p-component\\" data-pc-name=\\"inputtext\\" data-pc-section=\\"root\\" id=\\"names\\">
  </div>
  <div class=\\"field\\">
    <div class=\\"field-header\\"><label for=\\"lastnames\\">Apellidos</label>
      <!--v-if-->
    </div><input class=\\"p-inputtext p-component\\" data-pc-name=\\"inputtext\\" data-pc-section=\\"root\\" id=\\"lastnames\\">
  </div>
  <div class=\\"field\\">
    <div class=\\"field-header\\"><label for=\\"email\\">Email</label>
      <!--v-if-->
    </div><input class=\\"p-inputtext p-component\\" data-pc-name=\\"inputtext\\" data-pc-section=\\"root\\" id=\\"email\\">
  </div>
  <div class=\\"field\\">
    <div class=\\"field-header\\"><label for=\\"cuit\\">CUIT</label>
      <!--v-if-->
    </div><input class=\\"p-inputtext p-component\\" data-pc-name=\\"inputtext\\" data-pc-section=\\"root\\" id=\\"cuit\\">
  </div>
  <div class=\\"field\\">
    <div class=\\"field-header\\"><label for=\\"birthdate\\">Fecha de nacimiento</label>
      <!--v-if-->
    </div><input class=\\"p-inputtext p-component\\" data-pc-name=\\"inputtext\\" data-pc-section=\\"root\\" id=\\"birthdate\\">
  </div>
  <div class=\\"field\\">
    <div class=\\"field-header\\"><label for=\\"residence\\">Residencia</label>
      <!--v-if-->
    </div><input class=\\"p-inputtext p-component\\" data-pc-name=\\"inputtext\\" data-pc-section=\\"root\\" id=\\"residence\\">
  </div>
  <div class=\\"field\\">
    <div class=\\"field-header\\"><label for=\\"cellphone\\">Celular</label>
      <!--v-if-->
    </div><input class=\\"p-inputtext p-component\\" data-pc-name=\\"inputtext\\" data-pc-section=\\"root\\" id=\\"cellphone\\">
  </div>
</div>"
`;
